src/main/java/io/cdap/plugin/spark/HierarchyToRelationalUtils.java

```diff
@@ -0,0 +1,211 @@
+    data.registerTempTable(SQLUtils.NAME_HIERARCHY_TABLE);
+    Dataset<Row> rootRecords = sqlContext.sql(SQLUtils.sqlStatementForRoot(config, SQLUtils.NAME_HIERARCHY_TABLE));
+    if (rootRecords.select(config.getParentField()).distinct().count() > 1) {
+      throw new Exception("There can only be one root element in the hierarchy");
+    }
+    Row root = rootRecords.select(config.getParentField()).distinct().first();
+
+    Dataset<Row> uniqueRecords = data.sqlContext().sql(
+      SQLUtils.sqlForUniqueRecords(config, root, SQLUtils.NAME_HIERARCHY_TABLE));
+    // Fetch on child id's from the input dataset
+    // TODO change hardcoded values
+    Dataset<Row> leafRecords = data.sqlContext().sql(
+      SQLUtils.sqlStatementForLeafRecords(config, SQLUtils.NAME_HIERARCHY_TABLE));
+    leafRecords.registerTempTable(SQLUtils.NAME_LEAVES_TEMP_TABLE);
+    // TODO - required in order to be able to iterate - can pass data and sqlContext otherwise
+    List<Row> rows = uniqueRecords.collectAsList();
+
+    // Empty Dataset to hold all the results
+    Dataset<Row> results = sqlContext.createDataFrame(
+      sparkExecutionPluginContext.getSparkContext().emptyRDD(), DataFrames.toDataType(outputSchema));
+    // Root does not exists as separate value - this will generate it
+    results = results.union(generateTreeRoot(data, config));
+    // Generate tree for each unique row
+    for (Row row : rows) {
+      results = results.union(generateQuery(row, data, sqlContext));
+    }
+    // sets leaves indicator
+    results.registerTempTable(SQLUtils.NAME_RESULTS_TEMP_TABLE);
+    return sqlContext.sql(SQLUtils.sqlStatementForSettingLeafsValue(config));
+  }
+
+  /**
+   * Generate dataset containing all subtree details for a given row
+   *
+   * @param parentRow  row for which to extract the subtree
+   * @param data       reference to input dataset
+   * @param sqlContext reference to {@link SQLContext}
+   * @return dataset reference with all subtree rows for a given row
+   * @throws Exception raises exception if max depth is exceeded
+   */
+  public Dataset<Row> generateQuery(Row parentRow, Dataset<Row> data, SQLContext sqlContext)
+    throws Exception {
+    long dataFrameCount = 1;
+    int count = 1;
+
+    // make parent id sql ready based on type
+    String parentId;
+    String parentIdValue = String.valueOf(parentRow.get(0));
+    if (NumberUtils.isNumber(parentIdValue)) {
+      parentId = parentIdValue;
+    } else {
+      parentId = String.format("'%s'", parentRow.getString(0));
+    }
+
+    // TODO - will require all input fields and config of parent fields in order to swap parent id and parent field
+    // ex: changing from 11,13,Milk,Chocolate Milk with 3,13,Dairy,Chocolate Milk
+    // currently we have only parent id
+    mapSelfToTreeRoot(parentRow, data, config);
+
+    //TODO - move to centralized location (waiting on reply)
+    String parentProductField = "ParentProduct";
```

Comment:
why are we hardcoding field names? shouldn't we get them from the config?

Reply:
this is related to a reply we are waiting on the mapping for this part will look like.

Reply:
Let's add configs for parent product and child product fields, so we don't need to hardcode field names.